add Admin annotation

wodka opened the pull request.

wodka commented:
This Functionality depends on https://github.com/schmittjoh/JMSDiExtraBundle/pull/209 to enable custom annotations

OskarStark commented on composer.json:
```diff
@@ -16,6 +16,12 @@
+    "repositories":[
+        {
+            "type":											"vcs",
+            "url":											"https://github.com/wodka/JMSDiExtraBundle"
```
can you remove these spaces?

wodka replied:
I will remove the repositories property once the pull request in jms/di-extra-bundle is through (it's just to get it testable now)

OskarStark commented:
some tests would be great

wodka commented:
to be honest I'm not sure how to test this here - even if it should be tested here.

there are tests for the Custom Annotations in the DiExtraBundle

OskarStark commented:
i would then create a test admin class and execute these tests, but with the real keys.

what do you think?

btw. saw the tests @DI Bundle

wodka commented:
@OskarStark I just added 3 tests + improved the group / label matching to use the same as the Admin class.

for now it also requires to have @Service from DI to pick up the admins - I'm working on a solution to get rid of that requirement...

rande commented:
I will prefer to have this class in an dedicated library too avoid licencing issue between Apache and GPL licence.

wodka commented:
@rande it's not modifying the other lib in any way - it "just" implements an interface.

According to the licence: 
```
      "Derivative Works" shall mean any work, whether in Source or Object
      form, that is based on (or derived from) the Work and for which the
      editorial revisions, annotations, elaborations, or other modifications
      represent, as a whole, an original work of authorship. For the purposes
      of this License, Derivative Works shall not include works that remain
      separable from, or merely link (or bind by name) to the interfaces of,
      the Work and Derivative Works thereof.
```

wodka commented:
+ the AdminExtractor implements JMS\TranslationBundle which is also licensed under the Apache Licence.

Where do you have the GPL Licence from? Sonata is under MIT as far as I can see

Soullivaneuh commented on Annotation/Admin.php:
```diff
@@ -0,0 +1,164 @@
+use Sonata\AdminBundle\Admin\Admin as AdminClass;
+
+/**
+ * use annotations to define admin classes.
```
`use` -> `Use`

Soullivaneuh commented on Annotation/Admin.php:
```diff
@@ -0,0 +1,164 @@
+class Admin implements MetadataProcessorInterface
+{
+    /**
+     * service id - autogenerated per default.
```
`service` -> `Service`

Soullivaneuh commented on Annotation/Admin.php:
```diff
@@ -0,0 +1,164 @@
+    public $id;
+
+    /**
+     * admin class.
```
`admin` -> `Admin`

Soullivaneuh commented on Annotation/Admin.php:
```diff
@@ -0,0 +1,164 @@
+     * admin class.
+     *
+     * @var string
+     * @Required
```
`@Required`?

Soullivaneuh commented:
Okay please uppercase firs letters on comments. :-)

Soullivaneuh commented on composer.json:
```diff
@@ -46,10 +46,12 @@
         "sonata-project/intl-bundle": "~2.1",
-        "symfony/phpunit-bridge": "~2.7|~3.0"
+        "symfony/phpunit-bridge": "~2.7|~3.0",
+        "jms/di-extra-bundle": "dev-master as 1.6"
```
Should be `~1.6`.

Maybe a `conflict` key should be added if not working with earlier versions.

wodka replied:
I can change that to ~1.6 once that version is released

wodka replied:
switched to 1.7 (was just released)

greg0ire replied:
Please group jms packages together (we should set sort-packages to true in our config).

wodka commented:
just rebased the pull request on master and updated the dependencies to jms-di-extra 1.7 (this was released a few hours ago)

can you check if this looks fine to you?

greg0ire commented:
[Please run make cs](http://91.121.3.21/job/sonata-ci-bundle-admin/810/console)

greg0ire commented on Annotation/Admin.php:
```diff
@@ -0,0 +1,172 @@
+        );
+
+        // Remove empty entries
+        $tag = array_filter($tag, function ($v) { return !is_null($v); });
```
Isn't `$tag = array_filter($tag):` close enough?

wodka replied:
array_filter will also remove entries with false, but we need those, ie show_in_dashboard

greg0ire replied:
Ok great. I think your test will fail if someone thinks like me and tries to "simplify" (meaning "break") your function, so no need for a comment here I think.

greg0ire commented on Annotation/Admin.php:
```diff
@@ -0,0 +1,172 @@
+                'setTranslationDomain',
+                array(
+                    $this->translationDomain,
+                ),
```
If you want to have a line shorter than 50 chars, I think `array($this->translationDomain),` is enough

wodka replied:
it's now a single line

greg0ire commented on Annotation/Admin.php:
```diff
@@ -0,0 +1,172 @@
+        }
+
+        if (empty($this->group) || empty($this->label)) {
+            throw new \RuntimeException(
```
If the exception can be detected just by looking at the code / configuration file, it is a `LogicException`, not a `RuntimeException` (see the php manual definition).

wodka replied:
done

greg0ire commented on Tests/DependencyInjection/Compiler/AnnotationCompilerPassTest.php:
```diff
@@ -0,0 +1,129 @@
+        $annotation->group = 'myGroup';
+        $annotation->label = 'myLabel';
+        $annotation->showInDashboard = false;
+        $annotation->translationDomain = 'OMG';
```
lol

greg0ire commented on Tests/DependencyInjection/Compiler/AnnotationCompilerPassTest.php:
```diff
@@ -0,0 +1,129 @@
+            $meta->methodCalls[0],
+            array(
+                'setTranslationDomain',
+                array(
```
Please use multiline notation when really needed.

greg0ire commented:
Please fixup the 3 last commits into the first one. The feature looks great to me, can you please document it?

greg0ire commented:
You forgot to change the exception classname in your test ;)

wodka commented:
all done :)

greg0ire commented on Resources/doc/reference/advanced_configuration.rst:
```diff
@@ -120,6 +120,35 @@ application's config file:
+
+.. code-block:: bash
+
+    composer install jms/di-extra-bundle
```
```
 composer help install
Usage:
  install [options] [--] [<packages>]...

Arguments:
  packages                       Should not be provided, use composer require instead to add a given package to composer.json.
```

Please do as advised, and recommend `composer require` instead.

wodka replied:
upps, to much npm lately... changed it

greg0ire commented:
Thanks a lot! Good to merge IMO.

OskarStark commented on Resources/doc/reference/advanced_configuration.rst:
```diff
@@ -120,6 +120,35 @@ application's config file:
 
+
+Using Annotations to define Admins
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
please extend the documentation for such a big feature, and create a seperate article for this.

all possible annotation needs to be there with some examples, otherwise no one can use this feature

thanks

wodka replied:
I just moved it to a separate page, no clue why sonataCI is failing now

greg0ire replied:
Strange. Are you able to run the tests on your setup? If yes, try again after a `composer update`

greg0ire replied:
Other MR fail with the same message : https://travis-ci.org/sonata-project/SonataAdminBundle/jobs/112540825

wodka replied:
"make test" works for me

greg0ire replied:
O_o even after `composer update` ?

greg0ire replied:
If yes, maybe it is something we do in `.travis.yml` that make this fail…

wodka replied:
it is broken with the compose update, checking why right now

greg0ire replied:
Thanks! Must be a version constraint that is too loose…

wodka replied:
it wasn't a big change: https://gist.github.com/wodka/95457be1ccaea2d0efd8

-> Sonata\AdminBundle\Tests\Command\GenerateAdminCommandTest is failing now all the time because container is never set


easy fix is to extend the setup: 
```php
        $kernel->expects($this->any())
            ->method('getContainer')
            ->will($this->returnValue($this->container));
```

wodka replied:
https://github.com/sonata-project/SonataAdminBundle/pull/3624 << I added the change here

greg0ire replied:
Would you please make a PR for that ? I opened #3622 for that…

greg0ire commented:
Please rebase on the latest master, and fixup thep "update documentation" commit with the previous one.

wodka commented:
done :)

greg0ire commented:
Please include the new doc in a table of contents : 

> WARNING: document isn't included in any toctree

wodka commented:
I added it to the advanced section

greg0ire commented:
Sensible choice :)

greg0ire commented:
Good to merge :) @OskarStark

wodka commented:
is there anything else for me to do? @OskarStark

wodka commented:
yeah :)
